File: src/main/java/pro/tuscan/domain/player/PlayerDetailsMapper.java
package pro.tuscan.domain.player;

import pro.tuscan.adapter.api.response.PlayerDetailsResponse;
import pro.tuscan.adapter.api.response.dto.BanInfo;
import pro.tuscan.adapter.api.response.dto.GameDetails;
import pro.tuscan.client.player.Ban;
import pro.tuscan.client.player.Membership;
import pro.tuscan.client.player.PlayerDetails;

import java.time.ZonedDateTime;
import java.util.List;

public class PlayerDetailsMapper {

    private PlayerDetailsMapper() {
    }

    public static PlayerDetailsResponse mapToPlayerDetailsResponse(PlayerDetails details) {
        return new PlayerDetailsResponse(details.getPlayerId(),
                                         details.getNickname(),
                                         mapToCsgoGameDetails(details),
                                         details.getAvatarUrl(),
                                         details.getCountry(),
                                         details.getMemberships()
                                                 .stream()
                                                 .findFirst()
                                                 .map(m -> Membership.valueOf(m.toUpperCase()))
                                                 .orElse(null),
                                         mapBanInfo(details.getBans()));
    }

    private static GameDetails mapToCsgoGameDetails(PlayerDetails details) {
        return details.hasCsgoGame() ? new GameDetails(details.getGames().getCsgo().getFaceitElo(),
                                                       details.getGames().getCsgo().getLevel(),
                                                       details.getGames().getCsgo().getRegion(),
                                                       details.getGames().getCsgo().getSteamId()) : null;
    }

    private static BanInfo mapBanInfo(List<Ban> bans) {
        boolean activeBan = bans.stream().anyMatch(ban -> ban.getStartsAt().isBefore(ZonedDateTime.now()));
        return new BanInfo(activeBan);
    }

}

File: src/main/java/pro/tuscan/domain/player/PlayerDetailsMapper.kt
package pro.tuscan.domain.player

import pro.tuscan.adapter.api.GameDetails
import pro.tuscan.adapter.api.Membership
import pro.tuscan.adapter.api.PlayerDetailsResponse
import pro.tuscan.client.player.PlayerDetails

object PlayerDetailsMapper {

    @JvmStatic
    fun mapToPlayerDetailsResponse(details: PlayerDetails): PlayerDetailsResponse =
            details.let {
                PlayerDetailsResponse(
                        it.playerId,
                        it.nickname,
                        mapToCsgoGameDetails(it),
                        it.avatarUrl,
                        it.country,
                        it.memberships.first().let { Membership.valueOf(it.toUpperCase()) })
            }

    private fun mapToCsgoGameDetails(details: PlayerDetails): GameDetails? =
            details.games.csgo
                    ?.let { GameDetails(it.faceitElo, it.level, it.region, it.steamId) }
}
